trunk/NDHMS/Routing/Reservoirs/Persistence_Level_Pool_Hybrid_Module_README.md

```diff
@@ -0,0 +1,113 @@
+described above including calling level pool release along with weighting and combining that release with the weighted observed
+discharge, and finally calling mass balance checks before returning the release/outflow back to the model.
+
+The module_persistence_levelpool_hybrid_parameters.F module defines and instantiates objects for a hybrid type reservoir's
+parameters/properties. Parameters holds static/unchanging variables that are set when the given reservoir object is
+initialized/instantiated.
+
+The module_persistence_levelpool_hybrid_state.F module defines and instantiates objects for a hybrid type reservoir's state.
+State holds and tracks dynamic/changing variables that are only relevant to the given hybrid reservoir object and not other
+modules or areas of the system.
+
+The module_persistence_levelpool_hybrid_tests.F module holds unit tests that test for all components of a hybrid reservoir
+are properly initialized.
+
+Within the Reservoirs directory, the module_read_timeslice_data.F module reads USGS timeslice files to get gage discharge
+values that will be used by reservoirs. Future development will either expand this module or replicate its functionality to
+another module to read in other forms of observation and forecast data. An observation lookback period is passed in to
+determine how far back in time from the current model time the module will look for timeslice files. The observation
+resolution determines the time increments the module will look back. For instance, a standard lookback period would be
+18 hours with an observation resolution of 15 minutes, where a model current time of 8:00 PM would search for timeslice
+files at every 15 minute increment between 2:00 AM and 8:00 PM that day. The module will first search for the most recent
+timeslice files and grab the discharge for a particular lake/reservoir if the gage quality standard is met at that time.
+If a gage discharge is missing or if the gage quality standard is not met for any particular lake/reservoir in the given
+timeslice file, the module will continue to look back at every observation resolution increment until either all
+lakes/reservoirs have a good quality discharge or the end of the lookback period is reached. The total lookback seconds
+from current model time that the discharge is read will also be returned.
+
+Also within the Reservoirs directory, the module_reservoir_utilities.F module contains multiple functions used by the Hybrid
+module. The modify_for_projected_storage function is called from the hybrid release function and modifies the outflow if a
+projected storage falls below the minimum or exceeds the maximum for the reservoir. There are also multiple functions used
+for reading variables from the reservoir persistence parameter and timeslice NetCDF files.
+
+### Testing
+
+To compile and run the unit tests, go to the Reservoir directory in the terminal and type
+
+```
+make test
```

Comment:
This fails for me in docker.
Cheyenne is down, but it dosent look like a docker problem.

Reply:
Could you try 'make' and then 'make test'? If it works b/c of that, I need to modify the doc.

Reply:
@jdmattern you can also add a dependency to the make target to ensure the rest of the module is compiled before the tests.

Reply:
I tried that, but it failed on `make test`. I did not look too hard at it.

Reply:
@jdmattern-noaa

Reply:
```
jamesmcc@cheyenne2[1016]:~/WRF_Hydro/wrf_hydro_nwm_public/trunk/NDHMS/Routing/Reservoirs> make clean
rm -f *.o
rm -f *.mod
rm -f reservoir_tests
make -C Level_Pool clean
make[1]: Entering directory '/glade/u/home/jamesmcc/WRF_Hydro/wrf_hydro_nwm_public/trunk/NDHMS/Routing/Reservoirs/Level_Pool'
rm -f *.o
rm -f *.mod
make[1]: Leaving directory '/glade/u/home/jamesmcc/WRF_Hydro/wrf_hydro_nwm_public/trunk/NDHMS/Routing/Reservoirs/Level_Pool'
make -C Persistence_Level_Pool_Hybrid clean
make[1]: Entering directory '/glade/u/home/jamesmcc/WRF_Hydro/wrf_hydro_nwm_public/trunk/NDHMS/Routing/Reservoirs/Persistence_Level_Pool_Hybrid'
rm -f *.o
rm -f *.mod
make[1]: Leaving directory '/glade/u/home/jamesmcc/WRF_Hydro/wrf_hydro_nwm_public/trunk/NDHMS/Routing/Reservoirs/Persistence_Level_Pool_Hybrid'
jamesmcc@cheyenne2[1016]:~/WRF_Hydro/wrf_hydro_nwm_public/trunk/NDHMS/Routing/Reservoirs> make
#Build each sub module then build the module that depends on all sub modules
mpif90 -O2 -g -w -c -ftz -align all -fno-alias -fp-model precise -FR -convert big_endian  -I./ -I ../../MPP -I ../MPP -I ../mod -I ../../mod -I/glade/u/apps/ch/opt/netcdf/4.6.3/intel/18.0.5//include module_reservoir_utilities.F
mpif90 -O2 -g -w -c -ftz -align all -fno-alias -fp-model precise -FR -convert big_endian  -I./ -I ../../MPP -I ../MPP -I ../mod -I ../../mod -I/glade/u/apps/ch/opt/netcdf/4.6.3/intel/18.0.5//include module_reservoir_base.F
mpif90 -O2 -g -w -c -ftz -align all -fno-alias -fp-model precise -FR -convert big_endian  -I./ -I ../../MPP -I ../MPP -I ../mod -I ../../mod -I/glade/u/apps/ch/opt/netcdf/4.6.3/intel/18.0.5//include module_reservoir_read_timeslice_data.F
ar -r ../../lib/libHYDRO.a module_reservoir_utilities.o
ar -r ../../lib/libHYDRO.a module_reservoir_base.o
ar -r ../../lib/libHYDRO.a module_reservoir_read_timeslice_data.o
cp *.mod ../../mod
#Build the modules
make -C Level_Pool
make[1]: Entering directory '/glade/u/home/jamesmcc/WRF_Hydro/wrf_hydro_nwm_public/trunk/NDHMS/Routing/Reservoirs/Level_Pool'
#Build each sub module then build the module that depends on all sub modules
mpif90 -O2 -g -w -c -ftz -align all -fno-alias -fp-model precise -FR -convert big_endian  -I ../../../MPP -I ../../../mod -I/glade/u/apps/ch/opt/netcdf/4.6.3/intel/18.0.5//include module_levelpool_parameters.F
mpif90 -O2 -g -w -c -ftz -align all -fno-alias -fp-model precise -FR -convert big_endian  -I ../../../MPP -I ../../../mod -I/glade/u/apps/ch/opt/netcdf/4.6.3/intel/18.0.5//include module_levelpool_state.F
mpif90 -O2 -g -w -c -ftz -align all -fno-alias -fp-model precise -FR -convert big_endian  -I ../../../MPP -I ../../../mod -I/glade/u/apps/ch/opt/netcdf/4.6.3/intel/18.0.5//include module_levelpool.F
mpif90 -O2 -g -w -c -ftz -align all -fno-alias -fp-model precise -FR -convert big_endian  -I ../../../MPP -I ../../../mod -I/glade/u/apps/ch/opt/netcdf/4.6.3/intel/18.0.5//include module_levelpool_tests.F
ar -r ../../../lib/libHYDRO.a module_levelpool_parameters.o
ar -r ../../../lib/libHYDRO.a module_levelpool_state.o
ar -r ../../../lib/libHYDRO.a module_levelpool.o
ar -r ../../../lib/libHYDRO.a module_levelpool_tests.o
cp *.mod ../../../mod
make[1]: Leaving directory '/glade/u/home/jamesmcc/WRF_Hydro/wrf_hydro_nwm_public/trunk/NDHMS/Routing/Reservoirs/Level_Pool'
make -C Persistence_Level_Pool_Hybrid
make[1]: Entering directory '/glade/u/home/jamesmcc/WRF_Hydro/wrf_hydro_nwm_public/trunk/NDHMS/Routing/Reservoirs/Persistence_Level_Pool_Hybrid'
#Build each sub module then build the module that depends on all sub modules
mpif90 -O2 -g -w -c -ftz -align all -fno-alias -fp-model precise -FR -convert big_endian  -I ../../../MPP -I ../../../mod -I/glade/u/apps/ch/opt/netcdf/4.6.3/intel/18.0.5//include module_persistence_levelpool_hybrid_parameters.F
mpif90 -O2 -g -w -c -ftz -align all -fno-alias -fp-model precise -FR -convert big_endian  -I ../../../MPP -I ../../../mod -I/glade/u/apps/ch/opt/netcdf/4.6.3/intel/18.0.5//include module_persistence_levelpool_hybrid_state.F
mpif90 -O2 -g -w -c -ftz -align all -fno-alias -fp-model precise -FR -convert big_endian  -I ../../../MPP -I ../../../mod -I/glade/u/apps/ch/opt/netcdf/4.6.3/intel/18.0.5//include module_persistence_levelpool_hybrid.F
mpif90 -O2 -g -w -c -ftz -align all -fno-alias -fp-model precise -FR -convert big_endian  -I ../../../MPP -I ../../../mod -I/glade/u/apps/ch/opt/netcdf/4.6.3/intel/18.0.5//include module_persistence_levelpool_hybrid_tests.F
ar -r ../../../lib/libHYDRO.a module_persistence_levelpool_hybrid_parameters.o
ar -r ../../../lib/libHYDRO.a module_persistence_levelpool_hybrid_state.o
ar -r ../../../lib/libHYDRO.a module_persistence_levelpool_hybrid.o
ar -r ../../../lib/libHYDRO.a module_persistence_levelpool_hybrid_tests.o
cp *.mod ../../../mod
make[1]: Leaving directory '/glade/u/home/jamesmcc/WRF_Hydro/wrf_hydro_nwm_public/trunk/NDHMS/Routing/Reservoirs/Persistence_Level_Pool_Hybrid'
jamesmcc@cheyenne2[1017]:~/WRF_Hydro/wrf_hydro_nwm_public/trunk/NDHMS/Routing/Reservoirs> make test
mpif90 -O2 -g -w -c -ftz -align all -fno-alias -fp-model precise -FR -convert big_endian  -I./ -I ../../MPP -I ../MPP -I ../mod -I ../../mod -I/glade/u/apps/ch/opt/netcdf/4.6.3/intel/18.0.5//include hydro_stop.F
mpif90 -O2 -g -w -c -ftz -align all -fno-alias -fp-model precise -FR -convert big_endian  -I./ -I ../../MPP -I ../MPP -I ../mod -I ../../mod -I/glade/u/apps/ch/opt/netcdf/4.6.3/intel/18.0.5//include module_reservoir_utilities.F
mpif90 -O2 -g -w -c -ftz -align all -fno-alias -fp-model precise -FR -convert big_endian  -I./ -I ../../MPP -I ../MPP -I ../mod -I ../../mod -I/glade/u/apps/ch/opt/netcdf/4.6.3/intel/18.0.5//include module_reservoir_base.F
mpif90 -O2 -g -w -c -ftz -align all -fno-alias -fp-model precise -FR -convert big_endian  -I./ -I ../../MPP -I ../MPP -I ../mod -I ../../mod -I/glade/u/apps/ch/opt/netcdf/4.6.3/intel/18.0.5//include module_reservoir_read_timeslice_data.F
make -C Level_Pool
make[1]: Entering directory '/glade/u/home/jamesmcc/WRF_Hydro/wrf_hydro_nwm_public/trunk/NDHMS/Routing/Reservoirs/Level_Pool'
#Build each sub module then build the module that depends on all sub modules
mpif90 -O2 -g -w -c -ftz -align all -fno-alias -fp-model precise -FR -convert big_endian  -I ../../../MPP -I ../../../mod -I/glade/u/apps/ch/opt/netcdf/4.6.3/intel/18.0.5//include module_levelpool_parameters.F
mpif90 -O2 -g -w -c -ftz -align all -fno-alias -fp-model precise -FR -convert big_endian  -I ../../../MPP -I ../../../mod -I/glade/u/apps/ch/opt/netcdf/4.6.3/intel/18.0.5//include module_levelpool_state.F
mpif90 -O2 -g -w -c -ftz -align all -fno-alias -fp-model precise -FR -convert big_endian  -I ../../../MPP -I ../../../mod -I/glade/u/apps/ch/opt/netcdf/4.6.3/intel/18.0.5//include module_levelpool.F
mpif90 -O2 -g -w -c -ftz -align all -fno-alias -fp-model precise -FR -convert big_endian  -I ../../../MPP -I ../../../mod -I/glade/u/apps/ch/opt/netcdf/4.6.3/intel/18.0.5//include module_levelpool_tests.F
ar -r ../../../lib/libHYDRO.a module_levelpool_parameters.o
ar -r ../../../lib/libHYDRO.a module_levelpool_state.o
ar -r ../../../lib/libHYDRO.a module_levelpool.o
ar -r ../../../lib/libHYDRO.a module_levelpool_tests.o
cp *.mod ../../../mod
make[1]: Leaving directory '/glade/u/home/jamesmcc/WRF_Hydro/wrf_hydro_nwm_public/trunk/NDHMS/Routing/Reservoirs/Level_Pool'
make -C Persistence_Level_Pool_Hybrid
make[1]: Entering directory '/glade/u/home/jamesmcc/WRF_Hydro/wrf_hydro_nwm_public/trunk/NDHMS/Routing/Reservoirs/Persistence_Level_Pool_Hybrid'
#Build each sub module then build the module that depends on all sub modules
mpif90 -O2 -g -w -c -ftz -align all -fno-alias -fp-model precise -FR -convert big_endian  -I ../../../MPP -I ../../../mod -I/glade/u/apps/ch/opt/netcdf/4.6.3/intel/18.0.5//include module_persistence_levelpool_hybrid_parameters.F
mpif90 -O2 -g -w -c -ftz -align all -fno-alias -fp-model precise -FR -convert big_endian  -I ../../../MPP -I ../../../mod -I/glade/u/apps/ch/opt/netcdf/4.6.3/intel/18.0.5//include module_persistence_levelpool_hybrid_state.F
mpif90 -O2 -g -w -c -ftz -align all -fno-alias -fp-model precise -FR -convert big_endian  -I ../../../MPP -I ../../../mod -I/glade/u/apps/ch/opt/netcdf/4.6.3/intel/18.0.5//include module_persistence_levelpool_hybrid.F
mpif90 -O2 -g -w -c -ftz -align all -fno-alias -fp-model precise -FR -convert big_endian  -I ../../../MPP -I ../../../mod -I/glade/u/apps/ch/opt/netcdf/4.6.3/intel/18.0.5//include module_persistence_levelpool_hybrid_tests.F
ar -r ../../../lib/libHYDRO.a module_persistence_levelpool_hybrid_parameters.o
ar -r ../../../lib/libHYDRO.a module_persistence_levelpool_hybrid_state.o
ar -r ../../../lib/libHYDRO.a module_persistence_levelpool_hybrid.o
ar -r ../../../lib/libHYDRO.a module_persistence_levelpool_hybrid_tests.o
cp *.mod ../../../mod
make[1]: Leaving directory '/glade/u/home/jamesmcc/WRF_Hydro/wrf_hydro_nwm_public/trunk/NDHMS/Routing/Reservoirs/Persistence_Level_Pool_Hybrid'
mpif90 -O2 -g -w -c -ftz -align all -fno-alias -fp-model precise -FR -convert big_endian  -I./ -I ../../MPP -I ../MPP -I ../mod -I ../../mod -I/glade/u/apps/ch/opt/netcdf/4.6.3/intel/18.0.5//include reservoir_tests.F
reservoir_tests.F(203): error #6404: This name does not have a type, and must have an explicit type.   [GETCWD]
        istat=getcwd(cwd)
--------------^
compilation aborted for reservoir_tests.F (code 1)
Makefile:31: recipe for target 'test' failed
make: [test] Error 1 (ignored)
mpif90 -L/glade/u/apps/ch/opt/netcdf/4.6.3/intel/18.0.5//lib -lnetcdff -lnetcdf -o reservoir_tests \
	hydro_stop.o \
	module_reservoir_utilities.o \
	module_reservoir_base.o \
	module_reservoir_read_timeslice_data.o \
	Level_Pool/module_levelpool_parameters.o \
	Level_Pool/module_levelpool_state.o \
	Level_Pool/module_levelpool_tests.o \
	Level_Pool/module_levelpool.o \
	Persistence_Level_Pool_Hybrid/module_persistence_levelpool_hybrid_parameters.o \
	Persistence_Level_Pool_Hybrid/module_persistence_levelpool_hybrid_state.o \
	Persistence_Level_Pool_Hybrid/module_persistence_levelpool_hybrid_tests.o \
	Persistence_Level_Pool_Hybrid/module_persistence_levelpool_hybrid.o \
	reservoir_tests.o
ifort: error #10236: File not found:  'reservoir_tests.o'
Makefile:31: recipe for target 'test' failed
make: [test] Error 1 (ignored)
```

Reply:
`getcwd` is a custom GNU Extension to Fortran that doesn't work with Intel

Reply:
Actually, Intel appears to have a version but it requires a `USE IFPORT` to use it

Reply:
I just realized a way around this call to get working dir and tested it. I will remove this.